src/components/Trade/hooks/useTradeAmounts.tsx

```diff
@@ -0,0 +1,228 @@
+    fees?: DisplayFeeData<KnownChainIds>
+  }
+
+  type SetTradeAmountsSynchronousArgs = {
+    sellAssetId?: AssetId
+    buyAssetId?: AssetId
+    amount?: string
+    action?: TradeAmountInputField
+  }
+  type SetTradeAmountsArgs = CalculateAmountsArgs
+
+  // Selectors
+  const selectedCurrencyToUsdRate = useAppSelector(selectFiatToUsdRate)
+
+  // Constants
+  const sellAssetFormState = sellTradeAsset?.asset
+  const buyAssetFormState = buyTradeAsset?.asset
+
+  const { getUsdRates, getTradeQuote } = swapperApi.endpoints
+  const assets = useSelector(selectAssets)
+
+  const setTradeAmounts = useCallback(
+    (args: SetTradeAmountsArgs) => {
+      const { cryptoSellAmount, cryptoBuyAmount, fiatSellAmount, fiatBuyAmount } =
+        calculateAmounts(args)
+      const buyTradeAssetAmount = fromBaseUnit(cryptoBuyAmount, args.buyAsset.precision)
+      const sellTradeAssetAmount = fromBaseUnit(cryptoSellAmount, args.sellAsset.precision)
+      setValue('fiatSellAmount', fiatSellAmount)
+      setValue('fiatBuyAmount', fiatBuyAmount)
+      setValue('buyTradeAsset.amount', buyTradeAssetAmount)
+      setValue('sellTradeAsset.amount', sellTradeAssetAmount)
+    },
+    [setValue],
+  )
+
+  // Use the existing fiat rates and quote without waiting for fresh data
+  const setTradeAmountsAsynchronous = useCallback(
+    (args: SetTradeAmountsAsynchronousArgs) => {
+      const amount = args.amount ?? amountFormState
+      const action = args.action ?? actionFormState
+      const buyAsset = args.buyAsset ?? buyAssetFormState
+      const sellAsset = args.sellAsset ?? sellAssetFormState
+      const buyAssetUsdRate = args.buyAssetUsdRate ?? buyAssetFiatRateFormState
+      const sellAssetUsdRate = args.sellAssetUsdRate ?? sellAssetFiatRateFormState
+      const fees = args.fees ?? feesFormState
+      const tradeFee = bnOrZero(fees?.tradeFee).div(bnOrZero(buyAssetUsdRate))
+      if (sellAsset && buyAsset && amount && action) {
+        setTradeAmounts({
+          amount,
+          action,
+          buyAsset,
+          sellAsset,
+          buyAssetUsdRate,
+          sellAssetUsdRate,
+          selectedCurrencyToUsdRate,
+          tradeFee,
+        })
+      }
+    },
+    [
+      amountFormState,
+      actionFormState,
+      buyAssetFormState,
+      sellAssetFormState,
+      buyAssetFiatRateFormState,
+      sellAssetFiatRateFormState,
+      feesFormState,
+      setTradeAmounts,
+      selectedCurrencyToUsdRate,
+    ],
+  )
+
+  const getFirstAccountIdFromChainId = useCallback(
+    (chainId: ChainId) => {
+      const accountSpecifiers = accountSpecifiersList.find(specifiers => specifiers[chainId])
+      return accountSpecifiers?.[chainId]
+    },
+    [accountSpecifiersList],
+  )
+
+  // Use the args provided to get new fiat rates and a fresh quote
+  // Useful when changing assets where we expect response data to meaningfully change - so wait before updating amounts
+  const setTradeAmountsSynchronous = useCallback(
```

Comment:
I get what we mean here, that this is "blocking" from a user-flow perspective, but I'm not sure this is very clear given the `async` nature of that function and the fact it should be `await`ed.

Looking at the logic and this comment:

>  Eagerly update the input field to improve UX whilst we wait for API responses

I would be inclined to go with something along the lines of `optimistic` (UI update) or `suspense`

Reply:
You're right, this naming is confusing - updated in https://github.com/shapeshift/web/pull/2789/commits/5e2467f8891c5c0b1d2d85e660e04e0d76ed3373, to make it clearer what's actually going on.